### PR TITLE
tests: mock saveLhr and assert no unit test source changes

### PR DESCRIPTION
**Pull request by patrickhulce**
**Summary**
We wanted to *mock* the saveLhr function, not just spy it. Also adds the `git diff` test to our github unit test suite :)

**Related Issues/PRs**
fixes https://github.com/GoogleChrome/lighthouse/pull/11509/files#r499844389 and our travis builds

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -111,6 +111,9 @@ jobs:
     - run: yarn diff:sample-json
       if: matrix.os == 'windows-latest'
 
+    # Fail if any changes were written to source files (ex, from -GA).
+    - run: git diff --exit-code
```
Comment (connorjclark) [inline .github/workflows/ci.yml:115]: Wanna add to the smokes step too?

**Review by connorjclark — APPROVED** [review]

File: .github/workflows/ci.yml (hunk 1 above)
Comment (patrickhulce) [reply .github/workflows/ci.yml:115]: yeah I'll add everywhere I just noticed it happens on update:sample-json too

**Comment by patrickhulce** [comment]
the tests should now fail before I fix the update:sample-json path, making sure we're catching this now :)

**Comment by connorjclark** [comment]
LGTM

**Comment by patrickhulce** [comment]
soooooo @connorjclark after looking at [all the places affected](https://github.com/GoogleChrome/lighthouse/pull/11519/checks?check_run_id=1212175571#step:13:1) by the writing `lhr.json` change I'm not sure I like it as a thing. 

This PR changes the extension to `.report.json` for now, but why did we need it? The original PR doesn't have a description and -A creates a JSON already when you just add `--output=json` so I'm not super clear why we need this at all. Writing an audit output to a folder that's supposed to be the artifacts doesn't feel like the best possible solution (and rears its head here in [several places](https://github.com/GoogleChrome/lighthouse/pull/11519/checks?check_run_id=1212175571#step:13:1) as evidenced by the git clean failures in numerous tests that are difficult to mock). Is there another way we can work to solve whatever the root cause is that inspired #11519?

**Comment by connorjclark** [comment]
> soooooo @connorjclark after looking at all the places affected by the writing lhr.json change I'm not sure I like it as a thing.

I thought the mock you added would prevent this?

**Comment by patrickhulce** [comment]
> I thought the mock you added would prevent this?

It prevents it in those specific tests but apparently we exercise audit mode in several other places that weren't touched in #11509. 

Do you have any thoughts on the rest of my comment?

**Comment by connorjclark** [comment]
What if we save in the "artifact folder" if: -A is set and --output-path is json? That should avoid the test issues.

Currently it's too cumbersome to hold on to a LHR for a run, compared to how easy it is for artifacts. especially combined with your new script to save the latest-run folder.

**Comment by patrickhulce** [comment]
> What if we save in the "artifact folder" if: -A is set and --output includes json? That should avoid the test issues.

For future readers, @connorjclark discussed offline and settled on this solution with the addition of `-G` being used.

**Comment by brendankenny** [comment]
> For future readers, @connorjclark and I discussed offline and settled on this solution with the addition of `-G` being used.

amusing to look back at the original [`-GAR`](https://github.com/GoogleChrome/lighthouse/issues/1806) proposal thread

**Pull request merged**